# ML Project : Predicting Movie Revenue
Made by:

Corentin Maillard 21306

Mourad Mettioui 195019

## A) Data Understanding and analysis

### 1) Load the Dataset and libraries

In [6]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

dataset_file_path = 'dataset.csv'
dataset_supplement_file_path = 'dataset_supplement.csv'
# transform the first collum into an index
dataset_data = pd.read_csv(dataset_file_path, index_col=0)
dataset_supplement_data = pd.read_csv(dataset_supplement_file_path, index_col=0)

dataset_supplement_data.drop(columns=['title'], inplace=True)
left = dataset_data.set_index(['id'])
right = dataset_supplement_data.set_index(['movie_id'])
data_combind = left.join(right)


### 2) Undersand the structure of the dataset

In [7]:
# examine the contents of the resultant Dataset using the head() command, which grabs the first five rows:
data_combind.head()

vote_count  vote_average                                     title  \
id                                                                           
19995        11800           7.2                                    Avatar   
285           4500           6.9  Pirates of the Caribbean: At World's End   
206647        4466           6.3                                   Spectre   
49026         9106           7.6                     The Dark Knight Rises   
49529         2124           6.1                               John Carter   

                                               tagline    status  \
id                                                                 
19995                      Enter the World of Pandora.  Released   
285     At the end of the world, the adventure begins.  Released   
206647                           A Plan No One Escapes  Released   
49026                                  The Legend Ends  Released   
49529             Lost in our world, found in another.  Released   

                                         spoken_languages  duration  \
id                                                                    
19995   [{"iso_639_1": "en", "name": "English"}, {"iso...     162.0   
285              [{"iso_639_1": "en", "name": "English"}]     169.0   
206647  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...     148.0   
49026            [{"iso_639_1": "en", "name": "English"}]     165.0   
49529            [{"iso_639_1": "en", "name": "English"}]     132.0   

             revenue     release  \
id                                 
19995   2.787965e+09  2009-12-10   
285     9.610000e+08  2007-05-19   
206647  8.806746e+08  2015-10-26   
49026   1.084939e+09  2012-07-16   
49529   2.841391e+08  2012-03-07   

                                  countries_of_production  ...  popularity  \
id                                                         ...               
19995   [{"iso_3166_1": "US", "name": "United States o...  ...  150.437577   
285     [{"iso_3166_1": "US", "name": "United States o...  ...  139.082615   
206647  [{"iso_3166_1": "GB", "name": "United Kingdom"...  ...  107.376788   
49026   [{"iso_3166_1": "US", "name": "United States o...  ...  112.312950   
49529   [{"iso_3166_1": "US", "name": "United States o...  ...   43.926995   

                                                 overview  \
id                                                          
19995   In the 22nd century, a paraplegic Marine is di...   
285     Captain Barbossa, long believed to be dead, ha...   
206647  A cryptic message from Bond’s past sends him o...   
49026   Following the death of District Attorney Harve...   
49529   John Carter is a war-weary, former military ca...   

                                  original_title original_language  \
id                                                                   
19995                                     Avatar                en   
285     Pirates of the Caribbean: At World's End                en   
206647                                   Spectre                en   
49026                      The Dark Knight Rises                en   
49529                                John Carter                en   

                                                 keywords  \
id                                                          
19995   [{"id": 1463, "name": "culture clash"}, {"id":...   
285     [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
206647  [{"id": 470, "name": "spy"}, {"id": 818, "name...   
49026   [{"id": 849, "name": "dc comics"}, {"id": 853,...   
49529   [{"id": 818, "name": "based on novel"}, {"id":...   

                                            homepage  \
id                                                     
19995                    http://www.avatarmovie.com/   
285     http://disney.go.com/disneypictures/pirates/   
206647   http://www.sonypictures.com/movies/spectre/   
49026             http://www.thedarkknightrises.com/   
495

We can see the different features which are:
- vote count
- vote average
- title
- tagline
- status
- spoken languages
- duration
- release
- countries_of_production
- production
- popularity
- overview
- original_title
- original_language
- keywords
- id
- homepage
- genres
- financial_investment
- actors
- production_crew

And we have the target : 
- revenue

In [8]:
# we are going to see if all the data that we have are complete
data_combind.info()
# we can also use the isnull().sum() to see directly how many data are missing
data_combind.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
Index: 4803 entries, 19995 to 25975
Data columns (total 21 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   vote_count               4803 non-null   int64  
 1   vote_average             4323 non-null   float64
 2   title                    4803 non-null   object 
 3   tagline                  3959 non-null   object 
 4   status                   4803 non-null   object 
 5   spoken_languages         4803 non-null   object 
 6   duration                 4801 non-null   float64
 7   revenue                  4755 non-null   float64
 8   release                  4802 non-null   object 
 9   countries_of_production  4803 non-null   object 
 10  production               4803 non-null   object 
 11  popularity               4803 non-null   float64
 12  overview                 4800 non-null   object 
 13  original_title           4803 non-null   object 
 14  original_language       

vote_count                    0
vote_average                480
title                         0
tagline                     844
status                        0
spoken_languages              0
duration                      2
revenue                      48
release                       1
countries_of_production       0
production                    0
popularity                    0
overview                      3
original_title                0
original_language           240
keywords                      0
homepage                   3091
genres                        0
financial_investment        384
actors                        0
production_crew               0
dtype: int64

We can see that they are only 1712 non-null homepage on 4803 entires, so we will discard this feature because they are too many missing datas.

For the other features, we can see if they are complete or not

In [9]:
# we are going to see if all the data are different or not
data_combind.nunique()

vote_count                 1609
vote_average                 70
title                      4800
tagline                    3944
status                        3
spoken_languages            544
duration                    156
revenue                    3268
release                    3280
countries_of_production     469
production                 3697
popularity                 4802
overview                   4800
original_title             4801
original_language            37
keywords                   4222
homepage                   1691
genres                     1175
financial_investment        412
actors                     4761
production_crew            4776
dtype: int64

In [10]:
# we will drop the id because it is not an information that will ever be seen by the custommer.
# So we can assume that it is not an important feature for our model.
data_combind.drop(columns=['id'], inplace=True)
data_combind.drop(columns=['homepage'], inplace=True)
data_combind.info()

KeyError: "['id'] not found in axis"

### 3) Perform Exploratory Data Analysis (EDA)

In [ ]:
data_combind.describe()

vote_count  vote_average     duration       revenue   popularity   
count   4803.000000   4323.000000  4801.000000  4.755000e+03  4803.000000  \
mean     690.217989      6.090354   106.875859  8.231486e+07    21.492301   
std     1234.585891      1.193315    22.611935  1.630872e+08    31.816650   
min        0.000000      0.000000     0.000000  0.000000e+00     0.000000   
25%       54.000000      5.600000    94.000000  0.000000e+00     4.668070   
50%      235.000000      6.200000   103.000000  1.918402e+07    12.921594   
75%      737.000000      6.800000   118.000000  9.311911e+07    28.313505   
max    13752.000000     10.000000   338.000000  2.787965e+09   875.581305   

       financial_investment  
count          4.419000e+03  
mean           2.898466e+07  
std            4.065526e+07  
min            0.000000e+00  
25%            7.500000e+05  
50%            1.480000e+07  
75%            4.000000e+07  
max            3.800000e+08

In [ ]:
sns.pairplot(dataset_data)  # Visualize pairwise relationships using a pairplot
plt.show()

We can see that that the revenue will be affected by:
- the financial investment
- the vote count

For the other features (vote average, duration, popularity), it is not really visible the impact they have an the revenue

We can see that the vote average is affected by:
- the vote count
- the financial investment

We can see a big problem, that is the feature that are not number cannot be used to make an EDA with the way I make it.

We will use the one hot encoding approach that generally perform best on:
- the country_of_production
- the spoken_languages
- genres
- status
- original_language

For the feature like the title, tagline, overview, original_title we know that they can have an influence on the renenue but they are to complexe to exploit so we will discard them

In [ ]:
data_combind.head()

# print(dataset_data.Dtype == 'object')
import json
lang_comp = data_combind.spoken_languages
# print(lang_comp)

lang = []
for item in lang_comp:
    langperligne = []
    # print(item)
    lst = json.loads(item)
    # print(lst)
    # lang.append(lst['iso_639_1'])
    for item in lst:
        langperligne.append(item['iso_639_1'])
    print(langperligne)

# print(langperligne)

print(data_combind.nunique())

['en', 'es']
['en']
['fr', 'en', 'es', 'it', 'de']
['en']
['en']
['en', 'fr']
['en']
['en']
['en']
['en']
['en', 'fr', 'de']
['en', 'es', 'it', 'fr', 'de']
['en', 'tr', 'el', 'zh']
['en']
['en']
['en']
['en']
['en', 'es']
['en']
['en']
['en']
['en', 'th']
['en']
['is', 'en', 'ru', 'fr']
['en']
['en', 'fr', 'de', 'sv', 'it', 'ru']
['ro', 'en', 'de', 'ru']
['en', 'th']
['en']
['en']
['en', 'ru', 'zh']
['en']
['en']
['en']
['en']
['en', 'es']
['en']
['en']
['en']
['en']
['en', 'ja', 'it', 'fr']
['en']
['en', 'es']
['en', 'it']
['en']
['en']
['en']
['en']
['en']
['en']
['en']
['en']
['en']
['en', 'de', 'ru']
['en']
['en']
['en']
['en']
['la', 'en', 'fr', 'ja', 'zh']
['en', 'fr', 'hi', 'it', 'la', 'zh', 'pt', 'ru', 'bo']
['en']
['en']
['en']
['en', 'de']
['en']
['en', 'zh']
['en']
['en']
['en', 'fa', 'ur', 'ar']
['en']
['en']
['en', 'zh', 'sa']
['en']
['en']
['en']
['en', 'de', 'es']
['en', 'fr', 'gd']
['en']
['en']
['en', 'ru', 'fr']
['en']
['en']
['en']
['en']
['en', 'ja']
['en']
['en']
[